File: backend/accounts/AccountManager.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export interface DebridAccount {
  provider: 'real_debrid' | 'all_debrid' | 'torbox' | 'custom';
  apiKey: string;
  username?: string;
  status: 'active' | 'invalid' | 'expired' | 'unverified';
  expiresAt?: string;
  points?: number;
}

export interface UnrestrictResult {
  success: boolean;
  directUrl?: string;
  fileName?: string;
  fileSize?: number;
  error?: string;
}

export class AccountManager {
  private dataDir: string;
  private accountsFile: string;
  private accounts: Record<string, DebridAccount> = {};

  constructor() {
    this.dataDir = path.resolve(__dirname, '../data');
    this.accountsFile = path.join(this.dataDir, 'accounts.json');
    this.loadAccounts();
  }

  private loadAccounts(): void {
    try {
      if (!fs.existsSync(this.dataDir)) {
        fs.mkdirSync(this.dataDir, { recursive: true });
      }
      if (fs.existsSync(this.accountsFile)) {
        const raw = fs.readFileSync(this.accountsFile, 'utf-8');
        this.accounts = JSON.parse(raw);
      }
    } catch (_) {
      this.accounts = {};
    }
  }

  private saveAccounts(): void {
    try {
      if (!fs.existsSync(this.dataDir)) {
        fs.mkdirSync(this.dataDir, { recursive: true });
      }
      fs.writeFileSync(this.accountsFile, JSON.stringify(this.accounts, null, 2), 'utf-8');
    } catch (_) {}
  }

  public setAccount(account: DebridAccount): void {
    this.accounts[account.provider] = account;
    this.saveAccounts();
  }

  public removeAccount(provider: string): void {
    delete this.accounts[provider];
    this.saveAccounts();
  }

  public getAccount(provider: string): DebridAccount | undefined {
    return this.accounts[provider];
  }

  public listAccounts(): DebridAccount[] {
    return Object.values(this.accounts).map((acc) => ({
      ...acc,
      apiKey: acc.apiKey ? `${acc.apiKey.substring(0, 4)}...${acc.apiKey.slice(-4)}` : ''
    }));
  }

  /**
   * Verify Real-Debrid API Key and update status
   */
  public async verifyRealDebrid(apiKey: string): Promise<{ success: boolean; username?: string; expiresAt?: string; error?: string }> {
    try {
      const res = await fetch('https://api.real-debrid.com/rest/1.0/user', {
        headers: { Authorization: `Bearer ${apiKey}` }
      });
      if (!res.ok) {
        return { success: false, error: `Real-Debrid verification failed: HTTP ${res.status}` };
      }
      const data = await res.json();
      const account: DebridAccount = {
        provider: 'real_debrid',
        apiKey,
        username: data.username,
        status: data.type === 'premium' ? 'active' : 'expired',
        expiresAt: data.expiration,
        points: data.points
      };
      this.setAccount(account);
      return {
        success: true,
        username: data.username,
        expiresAt: data.expiration
      };
    } catch (err: any) {
      return { success: false, error: err.message || 'Network error verifying Real-Debrid' };
    }
  }

  /**
   * Attempt to unrestrict a target URL using configured debrid accounts.
   */
  public async unrestrict(targetUrl: string): Promise<UnrestrictResult> {
    const rd = this.accounts['real_debrid'];
    if (rd && rd.apiKey) {
      try {
        const formData = new URLSearchParams();
        formData.append('link', targetUrl);

        const res = await fetch('https://api.real-debrid.com/rest/1.0/unrestrict/link', {
          method: 'POST',
          headers: {
            Authorization: `Bearer ${rd.apiKey}`,
            'Content-Type': 'application/x-www-form-urlencoded'
          },
          body: formData.toString()
        });

        if (res.ok) {
          const data = await res.json();
          if (data.download) {
            return {
              success: true,
              directUrl: data.download,
              fileName: data.filename,
              fileSize: data.filesize
            };
          }
        }
      } catch (_) {}
    }

    return { success: false, error: 'No active debrid resolver matched this link.' };
  }
}

export const accountManager = new AccountManager();

File: backend/accounts/account_manager.py
import os
import json
import httpx
from typing import Dict, Any, Optional, List

class AccountManager:
    def __init__(self):
        self.base_dir = os.path.dirname(os.path.abspath(__file__))
        self.data_dir = os.path.abspath(os.path.join(self.base_dir, '..', 'data'))
        self.accounts_file = os.path.join(self.data_dir, 'accounts.json')
        self.accounts: Dict[str, Any] = {}
        self.load_accounts()

    def load_accounts(self) -> None:
        try:
            if not os.path.exists(self.data_dir):
                os.makedirs(self.data_dir, exist_ok=True)
            if os.path.isfile(self.accounts_file):
                with open(self.accounts_file, 'r', encoding='utf-8') as f:
                    self.accounts = json.load(f)
            else:
                self.accounts = {}
        except Exception:
            self.accounts = {}

    def save_accounts(self) -> None:
        try:
            if not os.path.exists(self.data_dir):
                os.makedirs(self.data_dir, exist_ok=True)
            with open(self.accounts_file, 'w', encoding='utf-8') as f:
                json.dump(self.accounts, f, indent=2)
        except Exception:
            pass

    def set_account(self, account: Dict[str, Any]) -> None:
        provider = account.get('provider')
        if provider:
            self.accounts[provider] = account
            self.save_accounts()

    def remove_account(self, provider: str) -> None:
        if provider in self.accounts:
            del self.accounts[provider]
            self.save_accounts()

    def get_account(self, provider: str) -> Optional[Dict[str, Any]]:
        return self.accounts.get(provider)

    def list_accounts(self) -> List[Dict[str, Any]]:
        result = []
        for acc in self.accounts.values():
            safe_acc = dict(acc)
            api_key = safe_acc.get('apiKey', '')
            if api_key and len(api_key) > 8:
                safe_acc['apiKey'] = f"{api_key[:4]}...{api_key[-4:]}"
            result.append(safe_acc)
        return result

    async def verify_real_debrid(self, api_key: str) -> Dict[str, Any]:
        try:
            async with httpx.AsyncClient(timeout=8.0) as client:
                res = await client.get(
                    'https://api.real-debrid.com/rest/1.0/user',
                    headers={'Authorization': f'Bearer {api_key}'}
                )
                if res.status_code != 200:
                    return {'success': False, 'error': f'Real-Debrid verification failed: HTTP {res.status_code}'}
                data = res.json()
                account = {
                    'provider': 'real_debrid',
                    'apiKey': api_key,
                    'username': data.get('username'),
                    'status': 'active' if data.get('type') == 'premium' else 'expired',
                    'expiresAt': data.get('expiration'),
                    'points': data.get('points')
                }
                self.set_account(account)
                return {
                    'success': True,
                    'username': data.get('username'),
                    'expiresAt': data.get('expiration')
                }
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def unrestrict(self, link: str) -> Dict[str, Any]:
        acc = self.get_account('real_debrid')
        if not acc or not acc.get('apiKey') or acc.get('status') != 'active':
            return {'success': False, 'error': 'No active Real-Debrid account available.'}
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                res = await client.post(
                    'https://api.real-debrid.com/rest/1.0/unrestrict/link',
                    headers={'Authorization': f"Bearer {acc['apiKey']}"},
                    data={'link': link}
                )
                if res.status_code != 200:
                    return {'success': False, 'error': f'Debrid unrestrict failed: HTTP {res.status_code}'}
                data = res.json()
                return {
                    'success': True,
                    'directUrl': data.get('download'),
                    'fileName': data.get('filename'),
                    'fileSize': data.get('filesize')
                }
        except Exception as e:
            return {'success': False, 'error': str(e)}

account_manager = AccountManager()
